File: backend/src/routes/travel.ts
import express from 'express';
import { travelMeansService } from '../services/travelMeansService.js';
import { authenticateToken } from '../middleware/auth.js';

const router = express.Router();

router.post('/means', authenticateToken, async (req, res) => {
  try {
    const { startLocation, cities, startDate, totalDays, passengers, preferences } = req.body;

    if (!startLocation || !cities || !Array.isArray(cities) || cities.length === 0 || !startDate || !totalDays || !passengers) {
      return res.status(400).json({
        success: false,
        message: 'Missing required fields: startLocation, cities (array), startDate, totalDays, passengers',
      });
    }

    const travelMeans = await travelMeansService.calculateTravelMeans({
      startLocation,
      cities,
      startDate: new Date(startDate),
      totalDays: parseInt(totalDays),
      passengers: parseInt(passengers),
      preferences,
    });

    return res.json({ success: true, data: travelMeans });
  } catch (error) {
    console.error('Error calculating travel means:', error);
    return res.status(500).json({
      success: false,
      message: 'Failed to calculate travel means',
      error: error instanceof Error ? error.message : 'Unknown error',
    });
  }
});

router.post('/route', authenticateToken, async (req, res) => {
  try {
    const { from, to, departureDate, passengers, preferences } = req.body;

    if (!from || !to || !departureDate || !passengers) {
      return res.status(400).json({
        success: false,
        message: 'Missing required fields: from, to, departureDate, passengers',
      });
    }

    const route = await travelMeansService.getTravelMeansForRoute(
      from,
      to,
      new Date(departureDate),
      parseInt(passengers),
      preferences
    );

    return res.json({ success: true, data: route });
  } catch (error) {
    console.error('Error getting travel route:', error);
    return res.status(500).json({
      success: false,
      message: 'Failed to get travel route',
      error: error instanceof Error ? error.message : 'Unknown error',
    });
  }
});

router.post('/recommendations', authenticateToken, async (req, res) => {
  try {
    const { routes } = req.body;

    if (!routes || !Array.isArray(routes)) {
      return res.status(400).json({
        success: false,
        message: 'Routes array is required',
      });
    }

    const recommendations = [
      {
        type: 'COST_EFFECTIVE',
        routeIndices: [0, 1],
        description: 'Most budget-friendly options',
        estimatedSavings: 150,
      },
    ];

    return res.json({ success: true, data: { recommendations } });
  } catch (error) {
    console.error('Error generating travel recommendations:', error);
    return res.status(500).json({
      success: false,
      message: 'Failed to generate recommendations',
      error: error instanceof Error ? error.message : 'Unknown error',
    });
  }
});

export default router;

File: backend-python/app/routes/travel.py
"""Travel routes — travel means calculation."""

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel

from app.deps import get_current_user
from app.models import User
from app.services.travel_means import travel_means_service

router = APIRouter()


class MeansRequest(BaseModel):
    startLocation: str
    cities: list[str]
    startDate: str
    totalDays: int
    passengers: int
    preferences: dict | None = None


class RouteRequest(BaseModel):
    from_: str
    to: str
    departureDate: str
    passengers: int
    preferences: dict | None = None

    model_config = {"populate_by_name": True}


@router.post("/means")
async def calculate_means(
    req: MeansRequest,
    user: User = Depends(get_current_user),
):
    from datetime import datetime
    result = await travel_means_service.calculate_travel_means(
        start_location=req.startLocation,
        cities=req.cities,
        start_date=datetime.fromisoformat(req.startDate),
        total_days=req.totalDays,
        passengers=req.passengers,
        preferences=req.preferences,
    )
    return {"success": True, "data": result}


@router.post("/route")
async def get_route(
    req: RouteRequest,
    user: User = Depends(get_current_user),
):
    from datetime import datetime
    route = await travel_means_service.get_travel_means_for_route(
        from_location=req.from_,
        to=req.to,
        departure_date=datetime.fromisoformat(req.departureDate),
        passengers=req.passengers,
        preferences=req.preferences,
    )
    return {"success": True, "data": route}


@router.post("/recommendations")
async def recommendations(
    routes: list[dict],
    user: User = Depends(get_current_user),
):
    recommendations = [
        {
            "type": "COST_EFFECTIVE",
            "routeIndices": [0, 1],
            "description": "Most budget-friendly options",
            "estimatedSavings": 150,
        }
    ]
    return {"success": True, "data": {"recommendations": recommendations}}
